File: solved/unique-paths.ts
/**
 * https://leetcode.com/problems/unique-paths/
 */
export { uniquePaths };

function uniquePaths(m: number, n: number): number {
  const memo: Record<string, number> = {};

  function dfs(x: number, y: number) {
    if (x > m || y > n) return 0;
    if (x === m && y === n) return 1;
    const key = `${x},${y}`;
    if (key in memo) return memo[key];

    let total = 0;
    total += dfs(x + 1, y);
    total += dfs(x, y + 1);
    memo[key] = total;
    return total;
  }

  return dfs(1, 1);
}

File: returned/python/today4/unique_paths.py
# https://leetcode.com/problems/unique-paths/description/

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        count = 0
        def dfs(x: int, y: int) -> None:
            if x == m - 1 and y == n -1:
                nonlocal count
                count += 1
                return
            if x > m - 1 or y > n - 1:
                return
            dfs(x, y+1)
            dfs(x+1, y)
        dfs(0,0)
        return count
